File: arrays_and_strings/11_container_with_most_water/TS/containerWater.ts
export {};

function maxArea(tank: number[]): number {
  let left: number = 0;
  let right: number = tank.length - 1;
  let maxArea: number = 0;

  while (left < right) {
    const wall: number = Math.min(tank[left], tank[right]);
    const distance: number = right - left;

    const currentArea: number = wall * distance;
    maxArea = Math.max(currentArea, maxArea);
    if (tank[left] < tank[right]) {
      left++;
    } else if (tank[right] < tank[left]) {
      right--;
    } else {
        left++
    }
  }

  return maxArea;
}

const testCases: number[][] = [
  [1, 8, 6, 2, 5, 4, 8, 3, 7],
  [1, 1],
  [4, 3, 2, 1, 4],
  [1, 2, 1],
  [1, 2, 3, 4, 5],
  [5, 4, 3, 2, 1],
];

// for (const height of testCases) {
//   console.log(`Input:       ${JSON.stringify(height)}`);
//   console.log(`Output:      ${maxArea(height)}`);
//   console.log("-".repeat(35));
// }

const height = testCases[0];
console.log(`Input:       ${JSON.stringify(height)}`);
console.log(`Output:      ${maxArea(height)}`);
console.log("-".repeat(35));

File: arrays_and_strings/11_container_with_most_water/Python/container_water.py
from typing import List

class Solution:
    def maxArea(self, tank: List[int]) -> int:
        left: int = 0
        right: int = len(tank) - 1
        maxArea: int = 0
        
        while left < right:
            wall: int = min(tank[left], tank[right])
            distance: int = right - left
            
            currentArea = wall * distance
            maxArea = max(maxArea, currentArea)
            if tank[left] < tank[right]:
                left+=1
            else:
                right-=1
        return maxArea


if __name__ == "__main__":
    test_cases = [
        [1, 8, 6, 2, 5, 4, 8, 3, 7],  # → 49, LeetCode's own example
        [1, 1],  # → 1, minimum valid input
        [4, 3, 2, 1, 4],  # → 16, same height at both ends
        [1, 2, 1],  # → 2, peak in middle
        [1, 2, 3, 4, 5],  # → 6, increasing heights
        [5, 4, 3, 2, 1],  # → 6, decreasing heights
    ]

    for height in test_cases:
        print(f"Input:       {height}")
        print(f"Output:      {Solution().maxArea(height)}")
        print("-" * 35)
